### PR TITLE
test: port peering test from sharness to Go 

### PR DESCRIPTION
**Pull request by guseggert**
This is the slowest test in the sharness test suite, because it has
very long sleeps. It usually takes 2+ minutes to run.

This new impl runs all peering tests in about 20 seconds, since it
polls for conditions instead of sleeping, and runs the tests in
parallel.

This also has an additional test case for a peer that was never online
and then connects.

### CODE REVIEW

**Comment by guseggert** [comment]
This is ready to go, just waiting on https://github.com/ipfs/kubo/pull/9505 (which this builds on), and then I'll rebase this and set as "Ready".

File: test/cli/harness/harness.go [hunk 1]
```diff
@@ -188,3 +190,16 @@ func (h *Harness) Cleanup() {
 		log.Panicf("removing temp dir %s: %s", h.Dir, err)
 	}
 }
+
+// ExtractPeerID extracts a peer ID from the given multiaddr, and fatals if it does not contain a peer ID.
+func (h *Harness) ExtractPeerID(m multiaddr.Multiaddr) peer.ID {
+	peerIDStr, err := m.ValueForProtocol(multiaddr.P_P2P)
+	if err != nil {
+		panic(err)
+	}
```
Comment (Jorropo) [inline test/cli/harness/harness.go:199]: This will incorrectly select the relay p2p's peer id in relayed addresses.
That because it does left to right matching instead of right to left.
I guess this is an edge case, but I would do:
```suggestion
	var peerIDStr string
	multiaddr.ForEach(m, func(c multiaddr.Component) bool {
	  if c.Protocol() != multiaddr.P_P2P {
	    return true
	  }
	  
	  peerIDStr = c.Value()
	  
	  return true
	})
	
	if peerIDStr == "" {
	  panic(multiaddr.ErrProtocolNotFound)
	}
```
